pkg/ratelimiter/rate_limiter.go

```diff
@@ -103,25 +106,10 @@ func (rl *TokenBucketRateLimiter) fillAndReturnEntry(limitType LimitType, bucket
 	if isPriority {
 		multiplier = rl.PriorityMultiplier
 	}
-	// The locking strategy is adapted from the following blog post: https://misfra.me/optimizing-concurrent-map-access-in-go/
-	rl.mutex.RLock()
-	currentVal, exists := rl.buckets[bucket]
-	rl.mutex.RUnlock()
-	if !exists {
-		rl.mutex.Lock()
-		currentVal = &Entry{
-			tokens:   uint16(limit.MaxTokens * multiplier),
-			lastSeen: time.Now(),
-			mutex:    sync.Mutex{},
-		}
-		rl.buckets[bucket] = currentVal
-		rl.mutex.Unlock()
-
-		return currentVal
+	if entry := rl.buckets2.getAndRefill(bucket, limit, multiplier, false); entry != nil {
```

Comment:
While this looks pretty buttoned up in the implementation right now, it would be helpful to add some tests asserting that an entry is never in both buckets at the same time. 

That feels like the regression I'd be worried about if I was modifying this code.